src/devices/esp8266.c

```diff
@@ -0,0 +1,848 @@
+                                  get_client_ip_cb, cb);
+}
+
+/**
+ *
+ * @param cb The callback to be invoked when the method completes.
+ */
+static bool connect_cb(struct at_rsp *rsp, void *up)
+{
+        static const char *cmd_name = "connect_cb";
+        void (*cb)(bool, const int) = up;
+        bool status = at_ok(rsp);
+        int chan_id = -1;
+
+        if (!status) {
+                cmd_failure(cmd_name, NULL);
+                goto do_cb;
+        }
+
+        /*
+         * <chan_id>,<connection status>
+         * 0,CONNECT
+         */
+        char *toks[3];
+        const int tok_cnt = at_parse_rsp_line(rsp->msgs[0], toks,
+                                              ARRAY_LEN(toks));
+
+        if (tok_cnt != 2) {
+                cmd_failure(cmd_name, "Unexpected # of tokens in response");
+                status = false;
+                goto do_cb;
+        }
+        chan_id = atoi(toks[0]);
+
+do_cb:
+        if (cb)
+                cb(status, chan_id);
+
+        return false;
+}
+
+/**
+ * Opens a connection to a given destination.
+ * @param chan_id The channel ID to use.  0 - 4
+ * @param proto The Network protocol to use.
+ * @param dest_port The destination port to send data to.
+ * @param udp_port The source port from which data originates (UDP ONLY).
+ * @param udp_mode Tells the device whether or not the destination of the
+ * UDP packets are allowed to change.  0 means it won't change.  1 means it
+ * will change up to 1 time.  2 means it can change any number of times.
+ * @return true if the request was queued, false otherwise.
+ */
+bool esp8266_connect(const int chan_id, const enum esp8266_net_proto proto,
+                     const char *ip_addr, const int dest_port,
+                     const int udp_port, const int udp_mode,
+                     void (*cb) (bool, const int))
+{
+        if (!check_initialized("connect"))
+                return false;
+
+        char cmd[64];
+        switch (proto) {
+        case ESP8266_NET_PROTO_TCP:
+                snprintf(cmd, ARRAY_LEN(cmd),
+                         "AT+CIPSTART=%d,\"TCP\",\"%s\",%d",
+                         chan_id, ip_addr, dest_port);
+                break;
+        case ESP8266_NET_PROTO_UDP:
+                snprintf(cmd, ARRAY_LEN(cmd),
+                         "AT+CIPSTART=%d,\"UDP\",\"%s\",%d,%d,%d",
+                         chan_id, ip_addr, dest_port, udp_port, udp_mode);
+                break;
+        }
+
+        return NULL != at_put_cmd(state.ati, cmd, _TIMEOUT_LONG_MS,
+                                  connect_cb, cb);
+}
+
+struct tx_info {
+        struct Serial *serial;
+        size_t len;
+        void (*cb)(int);
+};
+
+/**
+ * This call back is designed to handle two types of response (since that is the
+ * way this wonkey device works).  So based on the state and reply we figure out
+ * how to handle things.
+ */
+static bool send_data_cb(struct at_rsp *rsp, void *up)
```

Comment:
Its a callback method that gets called twice.  If you look at the switch you will see one fork does a `return true` at the end.  If during a callback you return `true`, then it implies there are more commands to be handled.  So it will wait for more replies to come from the modem and when it gets them and goes down one of the other paths, it will `return false`.  I'll clarify the docs.